doc/man/innconfval.1
.\" Automatically generated by Pod::Man 4.07 (Pod::Simple 3.32)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.if !\nF .nr F 0
.if \nF>0 \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    if !\nF==2 \{\
.        nr % 0
.        nr F 2
.    \}
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "INNCONFVAL 1"
.TH INNCONFVAL 1 "2015-09-12" "INN 2.6.3" "InterNetNews Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
innconfval \- Get configuration parameters from inn.conf
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBinnconfval\fR [\fB\-pstv\fR] [\fB\-i\fR \fIfile\fR] [\fIparameter\fR ...]
.PP
\&\fBinnconfval\fR \fB\-C\fR [\fB\-i\fR \fIfile\fR]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBinnconfval\fR normally prints the values of the parameters specified on
the command line.  By default, it just prints the parameter values, but if
\&\fB\-p\fR, \fB\-s\fR, or \fB\-t\fR are given, it instead prints the parameter and
value in the form of a variable assignment in Perl, Bourne shell, or Tcl
respectively.  If no parameters are specifically requested, \fBinnconfval\fR
prints out all parameter values (this isn't particularly useful unless one
of \fB\-p\fR, \fB\-s\fR, or \fB\-t\fR were specified).
.PP
All parameters are taken from \fIinn.conf\fR except for \fIversion\fR, which is
always the version string of \s-1INN.\s0
.PP
If given the \fB\-C\fR option, \fBinnconfval\fR instead checks \fIinn.conf\fR,
reporting any problems found to standard error.  \fBinnconfval\fR will exit
with status 0 if no problems are found and with status 1 otherwise.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-C\fR" 4
.IX Item "-C"
Check \fIinn.conf\fR rather than printing out the values of parameters.
.IP "\fB\-i\fR \fIfile\fR" 4
.IX Item "-i file"
Use \fIfile\fR as the source configuration file rather than \fIinn.conf\fR.
\&\fIfile\fR must be a valid \fIinn.conf\fR file and will be parsed the same as
\&\fIinn.conf\fR would be.
.IP "\fB\-p\fR" 4
.IX Item "-p"
Print out parameters as Perl assignment statements.  The variable name
will be the same as the \fIinn.conf\fR parameter, and string values will be
enclosed in single quotes with appropriate escaping.  Boolean values will
be mapped to the strings \f(CW\*(C`true\*(C'\fR or \f(CW\*(C`false\*(C'\fR.  List values will be mapped
to an array of strings.  \s-1NULL\s0 values are printed out with the \f(CW\*(C`undef\*(C'\fR value.
.Sp
Here is an example:
.Sp
.Vb 5
\&    $enableoverview = \*(Aqtrue\*(Aq;
\&    @extraoverviewadvertised = ( \*(AqNewsgroups\*(Aq, \*(AqInjection\-Info\*(Aq );
\&    @extraoverviewhidden = undef;
\&    $organization = \*(AqLet\e\*(Aqs try nasty "quotes"\*(Aq;
\&    $maxforks = 10;
.Ve
.Sp
If \fBinnconfval\fR is called via the Perl \f(CW\*(C`INN::Config\*(C'\fR module, all
these variables are properly exported.
.IP "\fB\-s\fR" 4
.IX Item "-s"
Print out parameters as Bourne shell assignment statements.  The variable
name will be the \fIinn.conf\fR parameter name in all capitals, and all
variables will be exported, if not \s-1NULL. \s0 String values will be enclosed in
single quotes with appropriate escaping, and boolean values will be mapped
to \f(CW\*(C`true\*(C'\fR or \f(CW\*(C`false\*(C'\fR.  List values will be mapped to a space-separated
string representing an array of strings (as Bourne shell does not recognize
arrays, contrary to several other shells, an array cannot be returned for
interoperability reasons).
.Sp
Here is an example:
.Sp
.Vb 4
\&    ENABLEOVERVIEW=true; export ENABLEOVERVIEW;
\&    EXTRAOVERVIEWADVERTISED=\*(Aq"Newsgroups" "Injection\-Info"\*(Aq; export EXTRAOVERVIEWADVERTISED;
\&    ORGANIZATION=\*(AqLet\*(Aq\e\*(Aq\*(Aqs try nasty "quotes"\*(Aq; export ORGANIZATION;
\&    MAXFORKS=10; export MAXFORKS;
.Ve
.IP "\fB\-t\fR" 4
.IX Item "-t"
Print out parameters as Tcl assignment statements.  The variable name will
be the same as the \fIinn.conf\fR parameter name but with \f(CW\*(C`inn_\*(C'\fR prepended,
and string variables will be escaped appropriately.  Boolean values will
be mapped to the strings \f(CW\*(C`true\*(C'\fR or \f(CW\*(C`false\*(C'\fR.  List values will be mapped
to an array of strings.  \s-1NULL\s0 values are not printed out.
.Sp
Here is an example:
.Sp
.Vb 4
\&    set inn_enableoverview "true"
\&    set inn_extraoverviewadvertised { "Newsgroups" "Injection\-Info" }
\&    set inn_organization "Let\*(Aqs try nasty \e"quotes\e""
\&    set inn_maxforks 10
.Ve
.IP "\fB\-v\fR" 4
.IX Item "-v"
Print \s-1INN\s0's version.  This is equivalent to \f(CW\*(C`innconfval version\*(C'\fR.
.SH "HISTORY"
.IX Header "HISTORY"
Written by Rich \f(CW$alz\fR <rsalz@uunet.uu.net> for InterNetNews.
.PP
\&\f(CW$Id:\fR innconfval.pod 9288 2011\-07\-22 23:08:57Z iulius $
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIinn.conf\fR\|(5), \fIINN::Config\fR\|(3pm).


NEWS
Changes in 2.6.3

    * Fixed the selection of the elliptic curve to use with OpenSSL 1.1.0 or
      later; NIST P-256 was enforced instead of using the most secure curve.

    * A new inn.conf parameter has been added to fine-tune the cipher suites
      to use with TLS 1.3:  the *tlsciphers13* now permits configuring them.
      A separate cipher suite configuration parameter is needed for TLS 1.3
      because TLS 1.3 cipher suites are not compatible with TLS 1.2, and
      vice-versa.  In order to avoid issues where legacy TLS 1.2 cipher
      suite configuration configured in the *tlsciphers* parameter would
      inadvertently disable all TLS 1.3 cipher suites, the inn.conf
      configuration has been separated out.

    * Fixed a regression since INN 2.6.1 that prevented articles with
      internationalized header fields (that is to say encoded in UTF-8) from
      being posted.

    * Support for Python 3 has been added to INN.  Embedded Python filtering
      and authentication hooks for innd and nnrpd can now use version 3.3.0
      or later of the Python interpreter.  In the 2.x series, version 2.3.0
      or later is still supported.

      When configuring INN with the --with-python flag, the "PYTHON"
      environment variable, when set, is used to select the interpreter to
      embed.  Otherwise, it is searched in standard paths.

      In case you change the Python interpreter to embed, make sure that the
      Python scripts you use are written in the expected syntax for that
      version of the Python interpreter.  Notably, buffer objects have been
      replaced with memoryview objects in Python 3, and UTF-8 encoding now
      really matters for string literals (Python 3 uses bytes and Unicode
      objects).

      INN documentation and samples of Python hooks have been updated to
      provide more examples.

    * When a Python or Perl filter hook rejects an article, innd now
      mentions the reason in response to CHECK and TAKETHIS commands. 
      Previously, the reason was given only for the IHAVE command.

    * nnrpd now properly logs the hostname of clients whose connection
      failed owing to an issue during the negotiation of a TLS session or
      high load average.

Changes in 2.6.2

    * A new *syntaxchecks* parameter has been added in inn.conf.  It permits
      controlling the level of checks performed by innd and nnrpd.  Up to
      now, only one check can be enabled/disabled:  when *laxmid* is
      mentioned in the values of this new parameter, INN accepts Message-IDs
      that contain ".." in the left part, as well as Message-IDs with two
      "@" (such Message-IDs would otherwise be considered as syntactically
      invalid).  See the inn.conf(5) man page for more details.

      The check is disabled by default (*no-laxmid*), which corresponds to
      the legacy behaviour of INN 2.6.1 and earlier.

    * Use of the ovdb_server helper server is now the default when using the
      ovdb overview method, that is to say the default value for the
      *readserver* parameter in ovdb.conf is now set to true.  It improves
      stability and avoids deadlocks, timing issues and corrupted ovdb
      databases.

    * mailpost now removes empty header fields before attempting to post
      articles, and keeps trace of them in the X-Mailpost-Empty-Hdrs: newly
      generated header field body.  Also, mailpost now sanitizes header
      fields with regards to empty continuation header lines.  Thanks to
      Kamil Jonca for these bug reports.

    * A new -z parameter has been added to mailpost to mention a list of
      header fields to remove from the gated message.  Thanks to Dieter
      Stussy for the patch.

    * Fixed a bug in inews that was rejecting articles containing header
      fields whose length exceeded 998 bytes.  This limitation is for the
      length of a single line of a header field (and not for the length of
      the whole header field, as it was wrongly the case).

    * Added support for GnuPG's gpg binary (in addition to gpgv) in
      pgpverify.  Indeed, gpg still validates signatures made with weak
      digest algorithms like MD5 whereas gpgv no longer do.  Thanks to
      Thomas Hochstein for the patch, which permits validating control
      articles for hierarchies that are still using old PGP keys.

    * Added similar support for GnuPG's gpg binary in perl-nocem to validate
      NoCeM notices from issuers who are still using old PGP keys.

    * A few commands listed in the "Control commands to INND" section in
      daily Usenet reports were appearing as a mere letter; all of them are
      now properly converted to meaningful words.

    * The *tlsprotocols* parameter in inn.conf now recognizes the "TLSv1.3"
      value (for OpenSSL versions implementing TLS 1.3, that is to say
      starting from OpenSSL 1.1.1).

    * The buffindexed overview method will now hopefully work properly on
      systems with a native page size larger than 16KB.

    * Other minor bug fixes and documentation improvements.

Changes in 2.6.1

    * nnrpd now uses -0000 as the time zone for Date: and Injection-Date:
      header fields it generates.  It was previously using +0000, wrongly
      systematically indicating a local time zone at Universal Time when
      *localtime* is set to false (which is the default) in readers.conf. 
      The +0000 time zone will now be used only if *localtime* is set to
      true and UTC is really the local time zone of the server.

    * Julien Elie has implemented in nnrpd the new COMPRESS command
      described in draft-murchison-nntp-compress that extends the NNTP
      protocol to allow a connection to be effectively and efficiently
      compressed.  News clients that also support that extension will be
      able to benefit from that bandwidth optimization and improvement in
      speed.  Moreover, using COMPRESS is more secure than TLS-level

frontends/innconfval.c
/*  $Id: innconfval.c 7585 2006-11-21 09:37:51Z eagle $
**
**  Get a config value from INN.
*/

#include "config.h"
#include "clibrary.h"

#include "inn/innconf.h"
#include "inn/messages.h"
#include "inn/version.h"
#include "inn/libinn.h"

/*
**  Print the INN version string with appropriate quoting.
*/
static void
print_version(FILE *file, enum innconf_quoting quoting)
{
    switch (quoting) {
    case INNCONF_QUOTE_NONE:
        fprintf(file, "%s\n", INN_VERSION_STRING);
        break;
    case INNCONF_QUOTE_SHELL:
        fprintf(file, "VERSION='%s'; export VERSION\n", INN_VERSION_STRING);
        break;
    case INNCONF_QUOTE_PERL:
        fprintf(file, "$version = '%s';\n", INN_VERSION_STRING);
        break;
    case INNCONF_QUOTE_TCL:
        fprintf(file, "set inn_version \"%s\"\n", INN_VERSION_STRING);
        break;
    }
}


/*
**  Main routine.  Most of the real work is done by the innconf library
**  routines.
*/
int
main(int argc, char *argv[])
{
    int option, i;
    char *file = NULL;
    enum innconf_quoting quoting = INNCONF_QUOTE_NONE;
    bool okay = true;
    bool version = false;
    bool checking = false;

    message_program_name = "innconfval";

    while ((option = getopt(argc, argv, "Ci:pstv")) != EOF)
        switch (option) {
        default:
            die("usage error");
            break;
        case 'C':
            checking = true;
            break;
        case 'i':
            file = optarg;
            break;
        case 'p':
            quoting = INNCONF_QUOTE_PERL;
            break;
        case 's':
            quoting = INNCONF_QUOTE_SHELL;
            break;
        case 't':
            quoting = INNCONF_QUOTE_TCL;
            break;
        case 'v':
            version = true;
            break;
        }
    argc -= optind;
    argv += optind;

    if (version) {
        print_version(stdout, quoting);
        exit(0);
    }
    if (checking)
        exit(innconf_check(file) ? 0 : 1);

    /* Read in the inn.conf file specified. */
    if (!innconf_read(file))
        exit(1);

    /* Perform the specified action. */
    if (argv[0] == NULL) {
        innconf_dump(stdout, quoting);
        print_version(stdout, quoting);
    } else {
        for (i = 0; i < argc; i++)
            if (strcmp(argv[i], "version") == 0)
                print_version(stdout, quoting);
            else if (!innconf_print_value(stdout, argv[i], quoting))
                okay = false;
    }
    exit(okay ? 0 : 1);
}
